### PR TITLE
[WIP] Mkd add custom measurement

### PR DESCRIPTION
**Pull request by mdworken**
@jtu0 sorry for the big PR - I'd recommend viewing this as a commit journey though if feasible. I'm requesting you as a reviewer now, but I should still take another self-review pass when it's not 1:08 AM.

### CODE REVIEW

File: spec/fixtures/basic_custom_measurement.csv [hunk 1]
```diff
@@ -0,0 +1,7 @@
+measurement_event,entity,name,measurement,value,
```
Comment (mdworken) [inline spec/fixtures/basic_custom_measurement.csv:1]: @jtu0 Down the road we can consider expanding this to allow multiple `value`s to be specified to make creating bulk measurements cleaner and easier. Looking at White that may be highly desirable, but I don't think it's necessary for delivering the MVP to Pinto. Remind me to bring up this discussion when I meet with the stakeholders for White

File: app/models/measurement.rb [hunk 2]
```diff
@@ -1,3 +1,27 @@
 class Measurement < ApplicationRecord
-  belongs_to :tank
+  belongs_to :measurement_event
+
+  delegate :tank, to: :measurement_event
+  def self.create_from_csv_data(attrs)
+    # remove relational (non-attribute) data from hash to be handled separately
+    measurement_event_name = attrs.delete(:measurement_event)
+    tank = Tank.find_or_create_by!(name: attrs.delete(:name))
```
Comment (mdworken) [inline app/models/measurement.rb:8]: These `find_or_create_by!` calls freak me out in several ways, but they'll make our lives way easier for the demo. 

Down the road we should
- add a nice way to manage tanks/measurement events and not allow them to be autocreated in the import
and/or
- add a way to merge tanks or measurement events together if names are similar but not exact.

But that's a problem for later 😄

File: db/migrate/20200421024915_remove_measurments_from_tanks.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class RemoveMeasurmentsFromTanks < ActiveRecord::Migration[5.2]
```
Comment (jtu0) [inline db/migrate/20200421024915_remove_measurments_from_tanks.rb:1]: Probably not important? spelling on `Measurement`.
Relevant if we think code editors will pick up on it and autofill to misspellings.

File: spec/lib/csv_importer_spec.rb [hunk 4]
```diff
@@ -36,4 +36,25 @@
       end
     end
   end
+
+  describe "#process" do
+    let(:processed_file) { create(:processed_file) }
+    let(:category_name) { "Measurement" }
+    let(:file) { File.read(Rails.root.join("spec/fixtures/basic_custom_measurement.csv")) }
+    context "allows the upload of custom tank measurments (without notes)" do
```
Comment (jtu0) [inline spec/lib/csv_importer_spec.rb:44]: Probably not important? spelling on `Measurement`.
Relevant if we think code editors will pick up on it and autofill to misspellings.

File: db/seeds.rb [hunk 5]
```diff
@@ -47,10 +47,11 @@
...
-Measurement.create(name: 'count', value_type: 'integer', value: '743', tank: tank, date: 3.days.ago)
-Measurement.create(name: 'count', value_type: 'integer', value: '719', tank: tank, date: 1.days.ago)
+male = Animal.create!(sex: 'male')
+female = Animal.create!(sex: 'female')
+family = Family.create!(male: male, female: female)
+tank = Tank.create!(facility: Facility.find_by(code: 'PSRF'), name: 'AB-17')
+Operation.create!(tank: tank, animals_added: 800, animals_added_family_id: family.id, operation_date: 7.days.ago, operation_type: 'add')
+measurement_event = MeasurementEvent.create!(name: "My first measurement", tank: tank)
+Measurement.create!(name: 'count', value_type: 'integer', value: '743', measurement_event: measurement_event, date: 3.days.ago)
+Measurement.create!(name: 'count', value_type: 'integer', value: '719', measurement_event: measurement_event, date: 1.days.ago)
```
Comment (jtu0) [inline db/seeds.rb:57]: If these Measurements have the same MeasurementEvent, should they have the same date?

File: db/migrate/20200421024915_remove_measurments_from_tanks.rb (hunk 3 above)
Comment (mdworken) [reply db/migrate/20200421024915_remove_measurments_from_tanks.rb]: Fixed it, thanks!

**Pull request merged**